### ISSUE TITLE
RoPE weights missing from checkpoint. Were they present in training for MRL model?

### ISSUE THREAD

**Issue by joel-shor** [lbcb-sci/RiNALMo#38]
Can someone clarify whether RoPE was used in Rinalmo or not? I ask because I am getting errors of the form

`Missing key(s) in state_dict: lm.transformer.blocks.*.mh_attn.mh_attn.rotary_emb.inv_freq` when attempting to load the checkpoint weights on CPU. 

The manuscript and code defaults suggest that RoPE **was** used during training, but then I am confused as to why literally all the other weights are present in the checkpoints except for these.

1) Can an author confirm that the checkpoint weights were trained with a model using RoPE?
2) It's a stretch, but any chance one of the authors knows why these weights are NOT present in the checkpoint file, but expected for the CPU model?

**Comment by joel-shor** [comment]
Apologies, I've solved my own issue.

RoPE *WAS* used, and the correct way to load this model on CPU is to ignore the fact that there aren't values for the inv_freq register.

FWIW I have a working CPU version of your model, using a bit of PyTorch surgery. Let me know if this interests you.

Thanks for your time and attention.

**Comment by joel-shor** [comment]
FWIW, you can actually load the flash checkpoints in the current CPU branch by doing some simple checkpoint name remapping. The function to do the mapping is at the bottom of this comment. Results on the 7600 human 5' UTRs are the same as on your no flash branch, also below.

                 My custom attention     |       "no_flash" branch attention
                    Fidelity        GT MRL          Fidelity        GT MRL
                    R2      L1      R2      L1      R2      L1      R2      L1
    First 1K seqs   0.7700  0.5000                  0.7733  0.5072  0.6484  0.6190
    All 7600 seqs   0.7414  0.5128  0.6291  0.6187  0.7414  0.5128  0.6291  0.6187


def modify_flash_attn_wts_for_cpu(wts: dict[str, torch.Tensor]) -> dict[str, torch.Tensor]:
    """Modify flash attention weights for use on CPU.
    
    This means manually carving out regular attention from flash attention weights.
    
    General rules:
    - Map ex `lm.transformer.blocks.0.mh_attn.Wqkv.weight` ->
        'lm.transformer.blocks.0.mh_attn.mh_attn.to_q.weight',
        'lm.transformer.blocks.0.mh_attn.mh_attn.to_k.weight',
        'lm.transformer.blocks.0.mh_attn.mh_attn.to_v.weight',
    - Map ex `lm.transformer.blocks.0.mh_attn.out_proj.weight` ->
        'lm.transformer.blocks.0.mh_attn.mh_attn.out_proj.weight',
    """
    wts_keys = list(wts.keys())
    for k in wts_keys:
        if 'mh_attn.Wqkv.weight' in k:
            # Carve out 3 weights, and delete the flash attention weight.
            assert wts[k].ndim == 2
            assert wts[k].shape[0] % 3 == 0, wts[k].shape
            
            q_wts_name = k.replace('mh_attn.Wqkv.weight', 'mh_attn.mh_attn.to_q.weight')
            k_wts_name = k.replace('mh_attn.Wqkv.weight', 'mh_attn.mh_attn.to_k.weight')
            v_wts_name = k.replace('mh_attn.Wqkv.weight', 'mh_attn.mh_attn.to_v.weight')
            
            wts[q_wts_name], wts[k_wts_name], wts[v_wts_name]  = wts[k].chunk(3, dim=0)
            
            del wts[k]
        elif 'mh_attn.out_proj.weight' in k:
            # Map to the regular attention output projection weight.
            out_proj_wts_name = k.replace('mh_attn.out_proj.weight', 'mh_attn.mh_attn.out_proj.weight')
            wts[out_proj_wts_name] = wts[k]
            del wts[k]
            
    return wts
